src/Core/Echo.Ast/Construction/AstParser.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+using System.Collections.Generic;
+using Echo.ControlFlow;
+using Echo.ControlFlow.Regions;
+using Echo.ControlFlow.Serialization.Blocks;
+using Echo.DataFlow;
+
+namespace Echo.Ast.Construction
+{
+    /// <summary>
+    /// Transforms a <see cref="ControlFlowGraph{TInstruction}"/> and a <see cref="DataFlowGraph{TContents}"/> into an Ast
+    /// </summary>
+    public sealed class AstParser<TInstruction>
+    {
+        private readonly ControlFlowGraph<TInstruction> _controlFlowGraph;
+        private readonly AstArchitecture<TInstruction> _architecture;
+        private readonly BlockTransformer<TInstruction> _transformer;
+        private readonly Dictionary<BasicControlFlowRegion<TInstruction>, BasicControlFlowRegion<Statement<TInstruction>>> _regionsMapping =
+            new Dictionary<BasicControlFlowRegion<TInstruction>, BasicControlFlowRegion<Statement<TInstruction>>>();
+        
+        /// <summary>
+        /// Creates a new Ast parser with the given <see cref="ControlFlowGraph{TInstruction}"/>
+        /// </summary>
+        /// <param name="controlFlowGraph">The <see cref="ControlFlowGraph{TInstruction}"/> to parse</param>
+        /// <param name="dataFlowGraph">The <see cref="DataFlowGraph{TContents}"/> to parse</param>
+        public AstParser(ControlFlowGraph<TInstruction> controlFlowGraph, DataFlowGraph<TInstruction> dataFlowGraph)
+        {
+            _controlFlowGraph = controlFlowGraph;
+            _architecture = new AstArchitecture<TInstruction>(controlFlowGraph.Architecture);
+
+            var context = new AstParserContext<TInstruction>(_controlFlowGraph, dataFlowGraph);
```

Comment:
What's the point of a separate context class if `AstParser` contains mutable state as well? 

Probably best to just make `Parse` accept the CFG/DFG pair as parameters instead of the constructor. It can then create the context in there, store the CFG/DFG in the context only, and pass it around the other methods. This way the AST builder itself can stay stateless, with the exception of potential configuration properties that might be added in the future. This also allows for the reuse of `AstParser` instances.